### PR TITLE
Handle Marlin weight restoration and shape recording (QAT INT4 Rollout Part1)

### PR DESCRIPTION
**Pull request by GeLee-Q**
## Motivation

support Int4 RL weight transfer for Marlin-based models

## Modifications

1. Restore Marlin weights to their original format (e.g., GPTQ) before loading RL weight updates. This ensures updates are applied to the correct memory layout.
2. Explicitly record original shapes for MoE weights (w13, w2) to support rollback.
3. Prevent `_original_shapes` from being overwritten by Marlin shapes during tensor replacement.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

File: python/sglang/srt/layers/quantization/compressed_tensors/compressed_tensors_moe.py [hunk 1]
```diff
@@ -1016,13 +1016,38 @@ def create_weights(
...
+            # Only save if the key doesn't exist to prevent overwriting with Marlin shapes.
+            if name not in layer._original_shapes:
+                # This is a safety check; `create_weights` usually handles this already.
+                layer._original_shapes[name] = tuple(target_attr.shape)
+
             # It is important to use resize_() here since it ensures
             # the same buffer is reused
-            getattr(layer, name).resize_(new_t.shape)
-            getattr(layer, name).copy_(new_t)
+            target_attr.resize_(new_t.shape)
```
Comment (AniZpZ) [inline python/sglang/srt/layers/quantization/compressed_tensors/compressed_tensors_moe.py:1049]: any solution to ensure not oom here?

File: python/sglang/srt/layers/quantization/compressed_tensors/compressed_tensors_moe.py [hunk 2]
```diff
@@ -1103,7 +1128,23 @@ def replace_tensor(name, new_t):
             layer.w2_weight_scale.shape[2],
             self.group_size,
         )
-        replace_parameter(layer, "w2_weight_scale", marlin_w2_scales)
+        replace_tensor("w2_weight_scale", marlin_w2_scales)
+
+        layer.is_marlin_converted = True
+
+    def restore_weights_before_loading(self, layer: torch.nn.Module):
```
Comment (AniZpZ) [inline python/sglang/srt/layers/quantization/compressed_tensors/compressed_tensors_moe.py:1135]: this method only applied in https://github.com/sgl-project/sglang/pull/15245/changes and move it ?

**Review by AniZpZ — APPROVED** [review]

File: python/sglang/srt/layers/quantization/compressed_tensors/compressed_tensors_moe.py (hunk 1 above)
Comment (GeLee-Q) [reply python/sglang/srt/layers/quantization/compressed_tensors/compressed_tensors_moe.py:1049]: This code logic aligns with the original implementation and mirrors the `update_tensor_inplace` operation found in `replace_parameter.`  Additionally, the corresponding parameters are explicitly deleted after each update to prevent OOM. Assuming the previous code did not cause OOM issues, this implementation is expected to perform consistently.
